### ISSUE TITLE
Patch Drush so that dependencies of base profiles behave the same as normal profiles

### ISSUE THREAD

**Issue by danepowell** [acquia/lightning#308]
Currently Lightning has a hard dependency on Views UI and Database Logging modules. This makes it impossible to uninstall these modules in a production environment running Lightning, which would really be best practice.

Is this something that can be overridden by a subprofile? If not, these dependencies should probably be removed.

**Comment by phenaproxima** [comment]
That sounds OK to me. We could make them dependencies of Lightning Dev so that we will always automatically install them for internal development. @balsama, what do you think?

**Comment by balsama** [comment]
You should be able to uninstall them. Profile dependencies are supposed to be special in that they are bypassed during the check on the module uninstall page.

This previously _was_ a problem with the sub-profile patch. Then it was fixed. Now it appears to be back again. I think we need to fix the problem with the patch. And I think we need to prioritize it - hopefully before 2.0.6 next Wednesday.

**Comment by balsama** [comment]
See comment [256](https://www.drupal.org/node/1356276#comment-11819462) and then [262](https://www.drupal.org/node/1356276#comment-11886317)

**Comment by danepowell** [comment]
Ah your comment about profile dependencies being special makes sense. I was confused too, because I've never run into this problem in the past when uninstalling via the UI. But in this case, I was uninstalling via `core.extensions.yml`, and I got an error during the drupal install / config import process. It seems that the config import process doesn't carve out this same exception.

**Comment by balsama** [comment]
I'm without my computer til tomorrow and thus without phpstorm and a real dev environment. So I can't provide a new patch, but it looks to me like this logic was lost in the patch in comment 269.

```
-      foreach (array_keys($module->required_by) as $dependent) {
-        if ($dependent != $profile && drupal_get_installed_schema_version($dependent) != SCHEMA_UNINSTALLED) {
+      $dependents = array_diff_key($module->required_by, $profiles);
+      foreach (array_keys($dependents) as $dependent) {
+        if (drupal_get_installed_schema_version($dependent) != SCHEMA_UNINSTALLED) {
```

The second `-` line above had been:

`if (!in_array($dependent, array_keys($profiles)) && drupal_get_installed_schema_version($dependent) != SCHEMA_UNINSTALLED) {`

in the previous patch.

**Comment by balsama** [comment]
Actually, after doing some more testing, I can't reproduce this anymore. I wonder if this is happening because you got the most recent version of Lightning, but didn't get the core patch. (which is entirely possible given the idiosyncrasies with the composer-patches plugin)

**I was able to:**

_With Lightning as the profile:_
- Uninstall views_ui though the UI
- Uninstall views_ui through `drush config-import`
- Uninstall views_ui through `drush pm-uninstall views_ui`

_With a subprofile of Lightning as the profile:_
- Uninstall views_ui though the UI
- Uninstall views_ui through `drush config-import`

**What I wasn't able to do:**
_With a subprofile of Lightning as the profile:_
- Uninstall views_ui through `drush pm-uninstall views_ui`

**Comment by balsama** [comment]
(Ignore the most recent version of the core patch that I posted today. The testing above was done with 1356276-276-8.2.x.patch - the same one that shipped with Lightning 2.0.5)

**Comment by balsama** [comment]
@danepowell Can you confirm that you still see this problem. And, if so, what patches and version you have applied to core?

**Comment by mirsoftacquia** [comment]
I can confirm it's not possible to uninstall views_ui if using subprofile through `drush pm-uninstall views_ui`. This has implication on blt builds (among others), where you typically define to uninstall views_ui  in project.yml.
I think making them as dependencies of Lightning Dev would be much less error-prone in every case.

**Comment by balsama** [comment]
I agree that we need to fix the drush integration, but I'm trying to understand how this is currently breaking BLT builds.

Does BLT disable modules via drush? Or is there some other problem?

**Comment by mirsoftacquia** [comment]
Yes, BLT uninstalls modules defined in modules.${environment}.uninstall in `project.yml` file.

**Comment by mirsoftacquia** [comment]
And yes, it is using drush command. (vendor/acquia/blt/phing/tasks/setup.xml)

**Comment by balsama** [comment]
Updated Issue Title.

"Dependencies" is a bit of a misnomer when it comes to Profiles. They are really just enabled at install. There is no "hard" dependency as Drupal has always explicitly excluded dependency's of profiles when uninstalling a module.

Views UI and DB Log have always been dependencies of Lightning. They are also dependencies of Standard - so there is a precedent. Moving everything that someone _might_ want to uninstall to the dev module would result in Lightning having no dependencies at all - and a very unfriendly out of the box experience.

The problem here that I see is that Drush is only making the exception for the active profile. If you are using ProfileA as a sub-profile of Lightning, then ProfileA is the active profile and drush is treating Lightning as it would a module. The patch from #1356276 takes this into account and allows you to uninstall dependencies of the active profile and any other enabled parent profiles. We need to patch drush to do the same.

**Comment by mirsoftacquia** [comment]
@balsama Thanks of explanation, in that case it really makes sense to patch Drush rather than removing these dependencies.

**Comment by phenaproxima** [comment]
I created a hasty patch (not actually a patch, just replace drush_get_modules with this function) that should address this: https://gist.github.com/phenaproxima/a1bccfc78d551aa9371456750d31457c

**Comment by balsama** [comment]
PR against drush: https://github.com/drush-ops/drush/pull/2697

and Lightning: https://github.com/acquia/lightning/pull/311

**Comment by balsama** [comment]
Also a PR against the master branch of drush if anyone is using that: https://github.com/drush-ops/drush/pull/2698

**Comment by balsama** [comment]
Fixed in PR #308, commit 655518f3fdd91671bcfa84e26053a3ae901a37fa
